src/device_manager.rs

```diff
@@ -11,233 +11,287 @@
 //! vm_allocator to allocate the resources, ask vm_device to register the
 //! devices IO ranges, and finally set resources to virtual device.
 
-use crate::resources::Resource;
-use crate::{DeviceIo, IoAddress, IoSize};
-
-use std::cmp::{Ord, Ordering, PartialEq, PartialOrd};
-use std::collections::btree_map::BTreeMap;
-use std::result;
+use std::result::Result;
 use std::sync::Arc;
 
+use crate::bus::{self, BusManager, MmioAddress, MmioBus, MmioRange, PioAddress, PioBus, PioRange};
+use crate::resources::Resource;
+use crate::{DeviceMmio, DevicePio};
+
 /// Error type for `IoManager` usage.
 #[derive(Debug)]
 pub enum Error {
-    /// The inserting device overlaps with a current device.
-    DeviceOverlap,
-    /// The device doesn't exist.
-    NoDevice,
+    /// Error during bus operation.
+    Bus(bus::Error),
 }
 
-/// Simplify the `Result` type.
-pub type Result<T> = result::Result<T, Error>;
+/// Represents an object that provides PIO manager operations.
+pub trait PioManager {
+    /// Type of the objects that can be registered with this `PioManager`.
+    type D: DevicePio;
+
+    /// Return a reference to the device registered at `addr`, together with the associated
+    /// range, if available.
+    fn pio_device(&self, addr: PioAddress) -> Option<(&PioRange, &Self::D)>;
+
+    /// Dispatch a read operation to the device registered at `addr`.
+    fn pio_read(&self, addr: PioAddress, data: &mut [u8]) -> Result<(), bus::Error>;
+
+    /// Dispatch a write operation to the device registered at `addr`.
+    fn pio_write(&self, addr: PioAddress, data: &[u8]) -> Result<(), bus::Error>;
+
+    /// Register the provided device with the specified range.
+    fn register_pio(&mut self, range: PioRange, device: Self::D) -> Result<(), bus::Error>;
 
-// Structure describing an IO range.
-#[derive(Debug, Copy, Clone)]
-struct IoRange {
-    base: IoAddress,
-    size: IoSize,
+    /// Unregister the device currently registered at `addr` together with the
+    /// associated range.
+    fn unregister_pio(&mut self, addr: PioAddress) -> Option<(PioRange, Self::D)>;
```

Comment:
nit: s/unregister/deregister where it shows up

Reply:
Wow, never considered the distinction before. I think deregister indeed sounds better. I've added the change as a separate commit so it's easier to undo in case it's breaking something.